Source/Dafny/Compilers/Compiler-python.cs

```diff
@@ -1060,10 +1116,26 @@ protected class ClassWriter : IClassWriter {
           }
         case SpecialField sf: {
             GetSpecialFieldInfo(sf.SpecialId, sf.IdParam, objType, out var compiledName, out _, out _);
-            if (compiledName.Length > 0) {
-              compiledName = $".{(sf is ConstantField && internalAccess ? "_" : "")}{compiledName}";
-            }
-            return SuffixLvalue(obj, compiledName);
+            return SimpleLvalue(w => {
+              var customReceiver = NeedsCustomReceiver(sf) && sf.EnclosingClass is not TraitDecl;
+              if (customReceiver) {
+                w.Write(TypeName_Companion(objType, w, member.tok, member));
+              } else {
+                obj(w);
+              }
+              if (compiledName.Length > 0) {
+                w.Write($".{(sf is ConstantField && internalAccess ? "_" : "")}{compiledName}");
+              }
+              var sep = "(";
+              EmitTypeDescriptorsActuals(ForTypeDescriptors(typeArgs, member, false), member.tok, w, ref sep);
+              if (customReceiver) {
+                w.Write(sep + "self");
+                sep = ", ";
+              }
+              if (sep != "(") {
+                w.Write(")");
+              }
```

Comment:
I would extract this parenthesizing logic into a separate function

Reply:
I don't see a reasonably generic way. Those `sep`s are annoying to work with.